main/src/library/Fixpoint3/Phase/Stratifier.flix

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2021 Benjamin Dahse
+ * Copyright 2025 Casper Dalgaard Nielsen
+ *                Adam Yasser Tallouzi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+mod Fixpoint3.Phase.Stratifier {
+    use Fixpoint3.Ast.Datalog.BodyPredicate.BodyAtom
+    use Fixpoint3.Ast.Datalog.Constraint.Constraint
+    use Fixpoint3.Ast.Datalog.Datalog.{Datalog, Model, Join}
+    use Fixpoint3.Ast.Datalog.{Datalog, Constraint, HeadPredicate, BodyPredicate, Polarity, Fixity}
+    use Fixpoint3.Ast.Datalog.HeadPredicate.HeadAtom
+    use Fixpoint3.Ast.Shared.PredSym
+    use Fixpoint3.Util.getOrCrash
+    use Fixpoint3.PrecedenceGraph.{MutGraph, getPseudoStrata}
+
+    ///
+    /// Array allowing lookup of the id-part of `PredSym` to name-part of `PredSym`.
+    ///
+    type alias PredIdToName[r: Region] = Array[String, r]
+
+    ///
+    /// Computes the stratification of `d`. The result is a map from `PredSym` to pseudo
+    /// strata and a their position in the pseudo-stratum. The number is meaningless,
+    /// except for allowing to seperate `PredSym` in the same pseudo-stratum by which
+    /// strata they belong to to.
+    ///
+    @Internal
+    pub def stratify(d: Datalog): Map[PredSym, (Int32, Int32)] = region rc {
+        let graph = MutGraph.empty(rc);
+        mkDepGraphInternal(graph, d);
+        let scc = getPseudoStrata(graph);
+        let numberOfPredSyms = MutGraph.getSize(graph);
+        let predSyms = Array.empty(rc, numberOfPredSyms);
+        collectPredSym(predSyms, d);
+        let result = MutMap.empty(rc);
+        scc |> List.forEachWithIndex(pseudoStrat -> list ->
+            list |> List.forEachWithIndex(strat -> set ->
+                set |> Set.forEach(x -> {
+                    let predSym = PredSym.PredSym(Array.get(x, predSyms), Int32.toInt64(x));
+                    MutMap.put(predSym, (pseudoStrat, strat), result)
+                })
+            )
+        );
+        MutMap.toMap(result)
+    }
+
+    ///
+    /// Insert edges into `graph` for the dependencies in the Datalog program `d`.
+    /// `A(x) :- B(x), A(x), not C(x).` inserts the edge `B -> A` and `C -> A` in `graph`.
+    ///
+    def mkDepGraphInternal(graph: MutGraph[r], d: Datalog): Unit \ r = match d {
+        case Datalog(_, rules) => Vector.forEach(precedenceHelper(graph), rules)
+        case Model(_) => ()
+        case Join(d1, d2) => mkDepGraphInternal(graph, d1); mkDepGraphInternal(graph, d2)
```

Comment:
afaik we don't have multiple statements on one line, I would add two newlines here